libs/vkd3d/command.c

```diff
@@ -6899,6 +6899,174 @@ static void d3d12_command_queue_execute(struct d3d12_command_queue *command_queu
     vkd3d_queue_release(command_queue->vkd3d_queue);
 }
 
+static void d3d12_command_queue_bind_sparse(struct d3d12_command_queue *command_queue,
+        enum vkd3d_sparse_memory_bind_mode mode, struct d3d12_resource *dst_resource,
+        struct d3d12_resource *src_resource, unsigned int count,
+        struct vkd3d_sparse_memory_bind *bind_infos)
+{
+    VkSparseImageOpaqueMemoryBindInfo opaque_info;
+    const struct vkd3d_vk_device_procs *vk_procs;
+    VkSparseImageMemoryBind *image_binds = NULL;
+    VkSparseBufferMemoryBindInfo buffer_info;
+    VkSparseMemoryBind *memory_binds = NULL;
+    VkSparseImageMemoryBindInfo image_info;
+    VkBindSparseInfo bind_sparse_info;
+    unsigned int first_packed_tile;
+    struct vkd3d_queue *queue;
+    VkDeviceMemory vk_memory;
+    VkDeviceSize vk_offset;
+    unsigned int i, j, k;
+    VkQueue vk_queue;
+    VkResult vr;
+
+    TRACE("queue %p, dst_resource %p, src_resource %p, count %u, bind_infos %p.\n",
+          command_queue, dst_resource, src_resource, count, bind_infos);
+
+    vk_procs = &command_queue->device->vk_procs;
+
+    bind_sparse_info.sType = VK_STRUCTURE_TYPE_BIND_SPARSE_INFO;
+    bind_sparse_info.pNext = NULL;
+    bind_sparse_info.waitSemaphoreCount = 0;
+    bind_sparse_info.pWaitSemaphores = NULL;
+    bind_sparse_info.bufferBindCount = 0;
+    bind_sparse_info.pBufferBinds = NULL;
+    bind_sparse_info.imageOpaqueBindCount = 0;
+    bind_sparse_info.pImageOpaqueBinds = NULL;
+    bind_sparse_info.imageBindCount = 0;
+    bind_sparse_info.pImageBinds = NULL;
+    bind_sparse_info.signalSemaphoreCount = 0;
+    bind_sparse_info.pSignalSemaphores = NULL;
+
+    first_packed_tile = dst_resource->sparse.tile_count;
+
+    if (d3d12_resource_is_buffer(dst_resource))
+    {
+        if (!(memory_binds = vkd3d_malloc(count * sizeof(*memory_binds))))
+        {
+            ERR("Failed to allocate sparse memory bind info.\n");
+            goto cleanup;
+        }
+
+        buffer_info.buffer = dst_resource->u.vk_buffer;
+        buffer_info.bindCount = count;
+        buffer_info.pBinds = memory_binds;
+
+        bind_sparse_info.bufferBindCount = 1;
+        bind_sparse_info.pBufferBinds = &buffer_info;
+    }
+    else
+    {
+        unsigned int opaque_bind_count = 0;
+        unsigned int image_bind_count = 0;
+
+        if (dst_resource->sparse.packed_mips.NumPackedMips)
+            first_packed_tile = dst_resource->sparse.packed_mips.StartTileIndexInOverallResource;
+
+        for (i = 0; i < count; i++)
+        {
+            const struct vkd3d_sparse_memory_bind *bind = &bind_infos[i];
+
+            if (bind->dst_tile < first_packed_tile)
+                image_bind_count++;
+            else
+                opaque_bind_count++;
+
+            if (!(memory_binds = vkd3d_malloc(opaque_bind_count * sizeof(*memory_binds))) ||
```

Comment:
This looks kinda sketchy. Why allocate in a loop with no frees?

Reply:
Good catch, this is supposed to happen outside the loop.

Reply:
Should be fixed.